README.md
Syncthing-GTK
=============

GTK3 &amp; Python based GUI and notification area icon for [Syncthing](https://github.com/syncthing/syncthing)

[![screenshot1](http://i.imgur.com/N36wmBM.png)](http://i.imgur.com/eX250tQ.png) [![screenshot2](http://i.imgur.com/43mmnC7.png)](http://i.imgur.com/RTRgRdC.png) [![screenshot3](http://i.imgur.com/KDBYekd.png)](http://i.imgur.com/OZ4xEeH.jpg)

##### Supported Syncthing features
- Everything what WebUI can display
- Adding / editing / deleting nodes
- Adding / editing / deleting repositories
- Restart / shutdown server
- Editing daemon settings

##### Additional features
- First run wizard for initial configuration
- Running Syncthing daemon in background
- Half-automatic setup for new nodes and repositories
- Nautilus (a.k.a. Files), Nemo and Caja integration
- Desktop notifications

##### Like what I'm doing?
[![Help me become filthy rich on Liberapay](https://img.shields.io/badge/Help%20me%20become%20filthy%20rich%20on-Liberapay-yellow.svg)](https://liberapay.com/kozec) <sup>or</sup> [![donate anything with PayPal](https://img.shields.io/badge/donate_anything_with-Paypal-blue.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=77DQD3L9K8RPU&lc=SK&item_name=kozec&item_number=scc&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donate_LG%2egif%3aNonHosted)

##### Packages
- Ubuntu, Debian, deb-based distros: in [OpenSUSE Build Service](http://software.opensuse.org/download.html?project=home%3Akozec&package=syncthing-gtk).
- Arch Linux: In [[community] repository](https://www.archlinux.org/packages/community/any/syncthing-gtk/)
- Fedora: search for `syncthing-gtk` in Software Center
- SUSE (and other rpm-based distros): in [OpenSUSE Build Service](http://software.opensuse.org/download.html?project=home%3Akozec&package=syncthing-gtk).
- Flatpak (distro-agnostic): in [Flathub](https://flathub.org/apps/details/me.kozec.syncthingtk)
- Windows: Get [latest installer from here](https://github.com/kozec/syncthing-gui/releases/latest), or use [the Chocolatey package](https://chocolatey.org/packages/syncthing-gtk).
- Or, in worst case scenario, download [latest tarball](https://github.com/kozec/syncthing-gui/releases/latest), extract it and run syncthing-gtk.py.

##### Dependencies
- Python 3, GTK 3.12 or newer, libnotify and [PyGObject](https://live.gnome.org/PyGObject)
- [python-gi-cairo](https://packages.debian.org/sid/python-gi-cairo),
[gir1.2-notify](https://packages.debian.org/sid/gir1.2-notify-0.7) and [gir1.2-rsvg](https://packages.debian.org/sid/gir1.2-rsvg-2.0) on debian based distros (included in PyGObject elsewhere)
- [python-dateutil](http://labix.org/python-dateutil) (Python 2 version)
- [python-bcrypt](https://pypi.python.org/pypi/bcrypt/2.0.0)
- [setuptools](https://pypi.python.org/pypi/setuptools)
- [psmisc](http://psmisc.sourceforge.net) (for the `killall` command)
- [Syncthing](https://github.com/syncthing/syncthing) v0.13 or newer

##### Optional Dependencies
- libnotify for desktop notifications.
- nautilus-python, nemo-python or caja-python for filemanager integration
- [this Gnome Shell extension](https://extensions.gnome.org/extension/615/appindicator-support/), if running Gnome Shell
- [gir1.2-appindicator3](https://packages.debian.org/sid/gir1.2-appindicator3-0.1) (part of [libappindicator](https://launchpad.net/libappindicator)), if running Gnome Shell or Unity

##### Windows Building Dependencies _(you don't need to install these just to **run** Syncthing-GTK)_
- Python for Windows 2.7
- [PyGObject for Windows](http://sourceforge.net/projects/pygobjectwin32/) with GTK3 enabled (tested with version 3.14.0)
- [python-dateutil](http://labix.org/python-dateutil) (Python 2 version)
- [Python for Windows Extensions](http://sourceforge.net/projects/pywin32/)
- [WMI](http://timgolden.me.uk/python/wmi/index.html)
- [NSIS2](http://nsis.sourceforge.net/NSIS_2) with NSISdl, [ZipDLL](http://nsis.sourceforge.net/ZipDLL_plug-in) and [FindProcDLL](http://forums.winamp.com/showpost.php?p=2777729&postcount=8) plugins (optional, for building installer)

##### Related links
- https://syncthing.net
- https://forum.syncthing.net/t/syncthing-gtk-gui-for-syncthing-now-with-inotify-support/709
- https://forum.syncthing.net/t/lxle-a-respin-of-lubuntu-now-has-syncthing-included-by-default/1392


debian/changelog
syncthing-gtk (0.9.4.4+ds+git20220108+9023143f8b93-1) unstable; urgency=medium

  * Cherry-pick patches from upstream pull requests.
  * Drop KDE 4 support.
  * Reshuffle menu items to better match GNOME HIG.
  * Add a list of folders to the indicator menu.
  * Use dh-sequence-python3.
  * Generate locales.
  * Generate and install upstream changelog.

 -- Andrej Shadura <andrewsh@debian.org>  Sat, 08 Jan 2022 19:48:03 +0100

syncthing-gtk (0.9.4.4+ds+git20201209+c46fbd8-1) unstable; urgency=medium

  [ Rowan Thorpe ]
  * Encode the password before applying bcrypt.
  * Argument to socket.socket.listen must be an integer.
    Closes: #976675.

  [ Andrej Shadura ]
  * Use Ayatana indicators.
  * Wrap and sort debian/control.
  * Bump debhelper from old 12 to 13.

 -- Andrej Shadura <andrewsh@debian.org>  Thu, 10 Dec 2020 10:00:02 +0100

syncthing-gtk (0.9.4.4+ds+git20200927+d09a2ef-3) unstable; urgency=medium

  * Rebuild with no changes to the source.

 -- Andrej Shadura <andrewsh@debian.org>  Thu, 05 Nov 2020 08:54:40 +0100

syncthing-gtk (0.9.4.4+ds+git20200927+d09a2ef-2) unstable; urgency=medium

  * Reintroduce the package to Debian.
  * Switch to a Debian fork based on a Python 3 port.
  * Update debian/watch.
  * Build with Python 3.

 -- Andrej Shadura <andrewsh@debian.org>  Wed, 04 Nov 2020 12:31:41 +0100

syncthing-gtk (0.9.4.4-2) unstable; urgency=medium

  [ Debian Janitor ]
  * Set upstream metadata fields: Repository-Browse.

 -- Jelmer Vernooĳ <jelmer@debian.org>  Sun, 15 Mar 2020 15:11:56 +0000

syncthing-gtk (0.9.4.4-1) unstable; urgency=medium

  * New upstream release.
  * Add suggests for nemo-python, python-nautilus or python-caja.
    Closes: #939576
  * Bump debhelper from old 11 to 12.
  * Remove obsolete fields Name from debian/upstream/metadata.

 -- Jelmer Vernooĳ <jelmer@debian.org>  Sat, 21 Sep 2019 21:10:56 +0000

syncthing-gtk (0.9.4.3-1) unstable; urgency=medium

  * Add Enhances: nautilus, nemo, caja. Closes: #913650
  * add gitlab ci configuration.
  * Add watch file.
  * New upstream version.

 -- Jelmer Vernooĳ <jelmer@debian.org>  Thu, 15 Nov 2018 02:19:43 +0000

syncthing-gtk (0.9.4.2-2) unstable; urgency=high

  [ Steve Langasek ]
  * Fix typo in dependencies: gir-1.2-glib-2.0 => gir1.2-glib2.0.
    Closes: #913771

 -- Jelmer Vernooĳ <jelmer@debian.org>  Thu, 15 Nov 2018 01:51:13 +0000

syncthing-gtk (0.9.4.2-1) unstable; urgency=low

  * Initial release. (Closes: #829561)

 -- Jelmer Vernooĳ <jelmer@debian.org>  Fri, 09 Nov 2018 20:56:34 +0000


syncthing_gtk/tools.py
#!/usr/bin/env python3
"""
Syncthing-GTK - tools

Various stuff that I don't care to fit anywhere else.
"""



from gettext import gettext as _
from base64 import b32decode
from datetime import tzinfo, timedelta
from subprocess import Popen
from dateutil import parser
import re, os, sys, random, string, platform, logging, shlex, gettext, __main__
log = logging.getLogger("tools.py")

IS_WINDOWS	= sys.platform in ('win32', 'win64')
IS_XP = IS_WINDOWS and platform.release() in ("XP", "2000", "2003")
IS_GNOME, IS_UNITY, IS_KDE, IS_CINNAMON = [False] * 4
IS_XFCE, IS_MATE, IS_I3, IS_LXQT = [False] * 4

if "XDG_CURRENT_DESKTOP" in os.environ:
	desktops = os.environ["XDG_CURRENT_DESKTOP"].split(":")
	IS_UNITY = ("Unity" in desktops)
	if not IS_UNITY:
		IS_GNOME = ("GNOME" in desktops) or ("GNOME-Flashback" in desktops) or ("GNOME-Fallback" in desktops)
	IS_KDE   = ("KDE" in desktops)
	IS_CINNAMON = ("X-Cinnamon" in desktops)
	IS_MATE = ("MATE" in desktops)
	IS_XFCE = ("XFCE" in desktops)
	IS_I3 = ("i3" in desktops)
	IS_LXQT = ("LXQt" in desktops)
if os.environ.get("DESKTOP_SESSION") == "gnome":
	# Fedora...
	IS_GNOME = True
if os.environ.get("DESKTOP_STARTUP_ID", "").startswith("i3/"):
	IS_I3 = True

LUHN_ALPHABET			= "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567" # Characters valid in device id
VERSION_NUMBER			= re.compile(r"^v?([0-9\.]*).*")
LOG_FORMAT				= "%(levelname)s %(name)-13s %(message)s"
GETTEXT_DOMAIN			= "syncthing-gtk" # used by "_" function
DESKTOP_FILE = """[Desktop Entry]
Name=%s
Exec=%s
Icon=%s
Comment=%s
X-GNOME-Autostart-enabled=true
Hidden=false
Type=Application
"""

portable_mode_enabled = False

if IS_WINDOWS:
	# On Windows, WMI and pywin32 libraries are reqired
	import wmi, winreg

def luhn_b32generate(s):
	"""
	Returns a check digit for the string s which should be composed of
	characters from LUHN_ALPHABET
	"""
	factor, sum, n = 1, 0, len(LUHN_ALPHABET)
	for c in s:
		try:
			codepoint = LUHN_ALPHABET.index(c)
		except ValueError:
			raise ValueError("Digit %s is not valid" % (c,))
		addend = factor * codepoint
		factor = 1 if factor == 2 else 2
		addend = (addend // n) + (addend % n)
		sum += addend
	remainder = sum % n
	checkcodepoint = (n - remainder) % n
	return LUHN_ALPHABET[checkcodepoint]

def check_device_id(nid):
	""" Returns True if device id is valid """
	# Based on nodeid.go
	nid = nid.strip("== \t").upper() \
		.replace("0", "O") \
		.replace("1", "I") \
		.replace("8", "B") \
		.replace("-", "") \
		.replace(" ", "")
	if len(nid) == 56:
		for i in range(0, 4):
			p = nid[i*14:((i+1)*14)-1]
			try:
				l = luhn_b32generate(p)
			except Exception as e:
				log.exception(e)
				return False
			g = "%s%s" % (p, l)
			if g != nid[i*14:(i+1)*14]:
				return False
		return True
	elif len(nid) == 52:
		try:
			b32decode("%s====" % (nid,))
			return True
		except Exception:
			return False
	else:
		# Wrong length
		return False

def sizeof_fmt(size):
	for x in ('B','kB','MB','GB','TB'):
		if size < 1024.0:
			if x in ('B', 'kB'):
				return "%3.0f %s" % (size, x)
			return "%3.2f %s" % (size, x)
		size /= 1024.0

def ints(s):
	""" Works as int(), but returns 0 for None, False and empty string """
	if s is None : return 0
	if s == False: return 0
	if hasattr(s, "__len__"):
		if len(s) == 0 : return 0
	return int(s)

def get_header(headers, key):
	"""
	Returns value of single header parsed from headers array or None
	if header is not found
	"""
	if not key.endswith(":"): key = "%s:" % (key,)
	for h in headers:
		if h.startswith(key):
			return h.split(" ", 1)[-1]
	return None

class Timezone(tzinfo):
	def __init__(self, hours, minutes):
		if hours >= 0:
			self.name = "+%s:%s" % (hours, minutes)
		else:
			self.name = "+%s:%s" % (hours, minutes)
		self.delta = timedelta(minutes=minutes, hours=hours)
	
	def __str__(self):
		return "<Timezone %s>" % (self.name,)
	
	def utcoffset(self, dt):
		return self.delta
	
	def tzname(self, dt):
		return self.name
	
	def dst(self, dt):
		return timedelta(0)

def parsetime(m):
	""" Parses time received from Syncthing daemon """
	try:
		return parser.parse(m)
	except ValueError:
		raise ValueError("Failed to parse '%s' as time" % m)

def parse_config_arguments(lst):
	"""
	Parses list of arguments and variables set in configuration
	Returns tuple of (variables_dict, prefix_arguments, arguments_list)
	"""
	vars, preargs, args = {}, [], []
	split = shlex.split(lst, False, False)
	args_target = preargs if "!" in split else args
	for i in split:
		if "=" in i and not i.startswith("-"):
			# Environment variable
			k, v = i.split("=", 1)
			vars[k] = v
			continue
		elif i == "!":
			args_target = args
		elif len(i.strip()) > 0:
			# Argument
			args_target.append(i.strip())
	return vars, preargs, args

def delta_to_string(d):
	"""
	Returns approximate, human-readable and potentially localized
	string from specified timedelta object
	"""
	# Negative time, 'some time ago'
	if d.days == -1:
		d = - d
		if d.seconds > 3600:
			return _("~%s hours ago") % (int(d.seconds / 3600),)
		if d.seconds > 60:
			return _("%s minutes ago") % (int(d.seconds / 60),)
		if d.seconds > 5:
			return _("%s seconds ago") % (d.seconds,)
		return _("just now")
	if d.days < -1:
		return _("%s days ago") % (-d.days,)

	# Positive time, 'in XY minutes'
	if d.days > 0:
		return _("in %s days") % (d.days,)
	if d.seconds > 3600:
		return _("~%s hours from now") % (int(d.seconds / 3600),)
	if d.seconds > 60:
		return _("%s minutes from now") % (int(d.seconds / 60),)
	if d.seconds > 5:
		return _("%s seconds from now") % (d.seconds,)
	return _("in a moment")

def init_logging():
	"""
	Initializes logging, sets custom logging format and adds one
	logging level with name and method to call.
	"""
	logging.basicConfig(format=LOG_FORMAT)
	# Rename levels
	logging.addLevelName(10, "D")	# Debug
	logging.addLevelName(20, "I")	# Info
	logging.addLevelName(30, "W")	# Warning
	logging.addLevelName(40, "E")	# Error
	# Create additional, "verbose" level
	logging.addLevelName(15, "V")	# Verbose
	# Add 'logging.verbose' method
	def verbose(self, msg, *args, **kwargs):
		return self.log(15, msg, *args, **kwargs)
	logging.Logger.verbose = verbose
	# Wrap Logger._log in something that can handle utf-8 exceptions
	old_log = logging.Logger._log
	def _log(self, level, msg, args, exc_info=None, extra=None):
		args = tuple([str(c) for c in args])
		msg = str(msg)
		old_log(self, level, msg, args, exc_info, extra)
	logging.Logger._log = _log

def make_portable():
	"""
	Set's IS_PORTABLE flag to True. Has to be called before
	everything else.
	"""
	global portable_mode_enabled
	log.warning("Portable mode enabled")
	portable_mode_enabled = True

def is_portable():
	""" Returns True after make_portable() is called. """
	global portable_mode_enabled
	return portable_mode_enabled

_localedir = None

def init_locale(localedir=None):
	"""
	Initializes gettext-related stuff
	"""
	global _localedir
	_localedir = localedir
	gettext.bindtextdomain(GETTEXT_DOMAIN, localedir)
	gettext.bind_textdomain_codeset(GETTEXT_DOMAIN, "utf-8")
	gettext.textdomain(GETTEXT_DOMAIN)

def get_locale_dir():
	"""
	Returns localedir passed to init_locale or None
	"""
	global _localedir
	return _localedir

def set_logging_level(verbose, debug):
	""" Sets logging level """
	logger = logging.getLogger()
	if debug:		# everything
		logger.setLevel(0)
	elif verbose:	# everything but debug
		logger.setLevel(11)
	else:			# INFO and worse
		logger.setLevel(20)
	if (debug or verbose) and IS_WINDOWS:
		# Windows executable has no console to output to, so output is
		# written to logfile as well
		import tempfile
		logfile = tempfile.NamedTemporaryFile(delete=False,
			prefix="Syncthing-GTK-",
			suffix=".log")
		logfile.close()
		h = logging.FileHandler(logfile.name)
		h.setFormatter(logging.Formatter(LOG_FORMAT))
		logging.getLogger().addHandler(h)

def check_daemon_running():
	""" Returns True if syncthing daemon is running """
	if not IS_WINDOWS:
		# Unix
		if not "USER" in os.environ:
			# Unlikely
			return False
		# signal 0 doesn't kill anything, but killall exits with 1 if
		# named process is not found
		p = Popen(["killall", "-u", os.environ["USER"], "-q", "-s", "0", "syncthing"])
		p.communicate()
		return p.returncode == 0
	else:
		# Windows
		if not "USERNAME" in os.environ:
			# Much more likely
			os.environ["USERNAME"] = ""
		proclist = wmi.WMI().ExecQuery('select * from Win32_Process where Name LIKE "syncthing.exe"')
		try:
			proclist = list(proclist)
			for p in proclist:
				p_user = p.ExecMethod_('GetOwner').Properties_('User').Value
				if p_user == os.environ["USERNAME"]:
					return True
		except Exception:
			# Can't get or parse list, something is horribly broken here
			return False
		return False

def parse_version(ver):
	"""
	Parses ver as version string, returning integer.
	Only first 6 components are recognized; If version string uses less
	than 6 components, it's zero-paded from right (1.0 -> 1.0.0.0.0.0).
	Maximum recognized value for component is 255.
	If version string includes non-numeric character, part of string
	starting with this character is discarded.
	If version string starts with 'v', 'v' is ignored.
	"""
	comps = VERSION_NUMBER.match(ver).group(1).split(".")
	if comps[0] == "":
		if ver == "unknown-dev":
			# Exception for non-tagged releases.
			# See https://github.com/syncthing/syncthing-gtk/issues/133
			return parse_version("v9999.99")
		# Not even single number in version string
		return 0
	while len(comps) < 6:
		comps.append("0")
	res = 0
	for i in range(0, 6):
		res += min(255, int(comps[i])) << ((5-i) * 8)
	return res

def compare_version(a, b):
	"""
	Parses a and b as version strings.
	Returns True, if a >= b
	Returns False, if b > a
	"""
	return parse_version(a) >= parse_version(b)

def get_config_dir():
	"""
	Returns ~/.config, %APPDATA% or whatever has user set as
	configuration directory.
	"""
	if is_portable():
		return os.environ["XDG_CONFIG_HOME"]
	if IS_WINDOWS and not IS_XP:
		try:
			from . import windows
			return windows.get_unicode_home()
		except Exception:
			pass
	from gi.repository import GLib
	confdir = GLib.get_user_config_dir()
	if confdir is None or IS_XP:
		if IS_WINDOWS:
			if "LOCALAPPDATA" in os.environ:
				# W7 and later
				confdir = os.environ["LOCALAPPDATA"]
			elif "APPDATA" in os.environ:
				# XP
				from ctypes import cdll
				os_encoding = 'cp' + str(cdll.kernel32.GetACP())
				confdir = os.environ["APPDATA"].decode(os_encoding)
			else:
				# 95? :D
				confdir = os.path.expanduser("~/.config")
		else:
			# Linux
			confdir = os.path.expanduser("~/.config")
	return confdir

get_install_path = None
if IS_WINDOWS:
	def _get_install_path():
		"""
		Returns installation path from registry.
		Available only on Windows
		"""
		if is_portable():
			return os.environ["XDG_CONFIG_HOME"]
		try:
			key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, "Software\\SyncthingGTK")
			path, keytype = winreg.QueryValueEx(key, "InstallPath")
			path = str(path)
			winreg.CloseKey(key)
			return path
		except WindowsError:
			# This is really shouldn't happen. Use executable path.
			os.path.dirname(sys.executable)
		
	get_install_path = _get_install_path

def get_executable():
	"""
	Returns filename of executable that was used to launch program.
	"""
	if IS_WINDOWS:
		return os.path.join(get_install_path(), "syncthing-gtk.exe")
	else:
		executable = __main__.__file__
		if not os.path.isabs(executable):
			cwd = os.getcwd()
			executable = os.path.normpath(os.path.join(cwd, executable))
		if executable.endswith(".py"):
			executable = "/usr/bin/env python3 %s" % (executable,)
		return executable

def is_ran_on_startup(program_name):
	"""
	Returns True if specified program is set to be ran on startup, either
	by XDG autostart or by windows registry.
	Only name (desktop filename or registry key) is checked.
	"""
	if IS_WINDOWS:
		# Check if there is value for application in ...\Run
		try:
			key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, "Software\\Microsoft\\Windows\\CurrentVersion\\Run")
			trash, keytype = winreg.QueryValueEx(key, program_name)
			winreg.CloseKey(key)
			return keytype in (winreg.REG_SZ, winreg.REG_EXPAND_SZ, winreg.REG_MULTI_SZ)
		except WindowsError:
			# Key not found
			return False
	else:
		# Check if there application.desktop file exists
		desktopfile = os.path.join(get_config_dir(), "autostart", "%s.desktop" % (program_name,))
		if not os.path.exists(desktopfile):
			return False
		# Check if desktop file is not marked as hidden
		# (stupid way, but should work)
		is_entry = False
		with open(desktopfile, "r") as f:
			for line in f.readlines():
				line = line.strip(" \r\t\n").lower()
				if line == "[desktop entry]":
					is_entry = True
					continue
				if "=" in line:
					key, value = line.split("=", 1)
					if key.strip(" ") == "hidden":
						if value.strip(" ") == "true":
							# Desktop file is 'hidden', i.e. disabled
							return False
		# File is present and not hidden - autostart is enabled
		return is_entry

def set_run_on_startup(enabled, program_name, executable, icon="", description=""):
	"""
	Sets or unsets program to be ran on startup, either by XDG autostart
	or by windows registry.
	'Description' parameter is ignored on Windows.
	Returns True on success.
	"""
	if is_ran_on_startup(program_name) == enabled:
		# Don't do anything if value is already set
		return
	if IS_WINDOWS:
		# Create/delete value for application in ...\Run
		key = winreg.OpenKey(winreg.HKEY_CURRENT_USER,
			"Software\\Microsoft\\Windows\\CurrentVersion\\Run",
			0, winreg.KEY_ALL_ACCESS)
		if enabled:
			winreg.SetValueEx(key, program_name, 0,
				winreg.REG_SZ, '"%s"' % (executable,))
		else:
			winreg.DeleteValue(key, program_name)
		winreg.CloseKey(key)
	else:
		# Create/delete application.desktop with provided values,
		# removing any hidding parameters
		desktopfile = os.path.join(get_config_dir(), "autostart", "%s.desktop" % (program_name,))
		if enabled:
			try:
				os.makedirs(os.path.join(get_config_dir(), "autostart"), mode=0o700)
			except Exception:
				# Already exists
				pass
			try:
				with open(desktopfile, "w", encoding='utf-8') as f:
					desktop_contents = DESKTOP_FILE % (program_name, executable, icon, description)
					f.write(desktop_contents)
			except Exception as e:
				# IO errors or out of disk space... Not really
				# expected, but may happen
				log.warning("Failed to create autostart entry: %s", e)
				return False
		else:
			try:
				if os.path.exists(desktopfile):
					os.unlink(desktopfile)
			except Exception as e:
				# IO or access error
				log.warning("Failed to remove autostart entry: %s", e)
				return False
	return True

def can_upgrade_binary(binary_path):
	"""
	Returns True if binary seems to be writable and placed in writable
	directory. Result may be wrong on Windows, but it's still more
	accurate that os.access, that responds with complete fabulation --
	https://mail.python.org/pipermail/python-list/2011-May/604395.html
	"""
	if IS_WINDOWS:
		# Try to open file in same directory. It's not good idea trying
		# to open very same file as Windows throws IOError if file is
		# already open somewhere else (i.e. it's binary that is running)
		try:
			path = binary_path + ".new"
			if os.path.exists(path):
				f = open(path, "r+b")
				f.close()
			else:
				f = open(path, "wb")
				f.close()
				os.unlink(path)
			# return Maybe
			return True
		except Exception as e:
			log.exception(e)
			return False
	else:
		# Life is just simpler on Unix
		if not os.access(binary_path, os.W_OK):
			return False
		path = os.path.split(binary_path)[0]
		if not os.access(path, os.W_OK):
			return False
		return True


def generate_folder_id():
	"""
	Returns new, randomly generated folder ID in a1bc2-x9y7z format
	"""
	return "-".join((
		("".join(random.choice(string.ascii_lowercase + string.digits) for _ in range(5)))
		for _ in range(2)
	))

def escape_html_entities(string):
	return string.replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;').replace('"', '&quot;').replace("'", '&#39;')
